File: Medium/Daily Temperatures/daily-temperatures.py
class Solution:
    def dailyTemperatures(self, temperatures: List[int]) -> List[int]:
        res = [0] * len(temperatures)
        stack = []
        debug = False

        for i in range(len(temperatures)):
            debug and print(stack)
            t = temperatures[i]
            if len(stack) == 0:
                stack.append(
                    (i, t)
                )
                continue
            if stack[-1][1] >= t:
                stack.append(
                    (i, t)
                )
                continue
            else:
                while stack and stack[-1][1] < t:
                    ix, _ = stack.pop()
                    res[ix] = i-ix
                stack.append(
                    (i, t)
                )

        return res

File: Medium/Daily Temperatures/daily-temperatures.cpp
class Solution {
public:
    vector<int> dailyTemperatures(vector<int>& temperatures) {
        vector<int> res(temperatures.size(), 0);
        stack<tuple<int, int>> stack;

        for (int i = 0; i < temperatures.size(); ++i)
        {
            int t = temperatures[i];

            if (stack.size() == 0)
            {
                stack.push(make_tuple(i, t));
                continue;
            }

            if (get<1>(stack.top()) > t)
                stack.push(make_tuple(i, t));
            else
            {
                while (!stack.empty() && get<1>(stack.top()) < t)
                {
                    int ix = get<0>(stack.top());
                    stack.pop();
                    res[ix] = i - ix;
                }
                stack.push(make_tuple(i, t));
            }
        }

        return res;
    }
};
